lib/libc/Kconfig

```diff
@@ -11,6 +11,14 @@ config REQUIRES_FULL_LIBC
 	  Helper symbol to indicate some feature requires a C library implementation
 	  with more functionality than what MINIMAL_LIBC provides
 
+config SUPPORT_MINIMAL_LIBC
+	bool
+	default y
+
+config SUPPORT_NEWLIB_LIBC
```

Comment:
Hmm, we have `TOOLCHAIN_HAS_NEWLIB` now but with this change we will also get `SUPPORT_NEWLIB_LIBC`. Do we need to differentiate them?

Reply:
Indeed. If we are going to add this, we should probably get rid of `TOOLCHAIN_HAS_NEWLIB`. Same goes for `HAS_NEWLIB_LIBC_NANO`.

IMHO, `TOOLCHAIN_HAS_NEWLIB` sounds more intuitive and in line with the existing naming convention, so maybe we could just rename it to `TOOLCHAIN_HAS_NEWLIB` and get rid of the existing CMake variable (i.e. make it fully Kconfig-derived and not reliant on the CMake scripts).

As for `SUPPORT_MINIMAL_LIBC`, `TOOLCHAIN_HAS_MINIMAL_LIBC` does not really make sense, so we could probably keep it as `TOOLCHAIN_SUPPORTS_MINIMAL_LIBC` or something like that.

Reply:
@evgeniy-paltsev  @stephanosio  Please note that the `TOOLCHAIN_HAS_NEWLIB` is a CMake variable not accessible in Kconfig.
The reason for the CMake variable can be seen here: https://github.com/zephyrproject-rtos/zephyr/pull/12393
It's intended for the test infrastructure to filter the tests based on the capabilities of the toolchain.

The `SUPPORT_NEWLIB_LIBC` is a Kconfig setting that ensures that a user cannot select `CONFIG_NEWLIB_LIBC=y` if the toolchain does not support it.
https://github.com/zephyrproject-rtos/zephyr/blob/2b91ebe16e25e5ce17270ea36b88fbb34f5f28ed/lib/libc/Kconfig#L27

Now, originally the reason for not having this is Kconfig was because @nashif wanted to avoid creating Kconfig for toolchains and instead obtain this knowledge through the CMake cache:
https://github.com/zephyrproject-rtos/zephyr/pull/12393#issuecomment-452748742
> I am not a big fan of creating a Kconfig layer for toolchains, toolchain configuration should only be cmake (toolchain file in cmake speak).

Now, that may have been true back then, but with growing number of toolchains and also toolchains coming with their own C libraries, I think we should reconsider that.

Therefore my proposal is to introduce:
- `SUPPORT_MINIMAL_LIBC`
- `SUPPORT_NEWLIB_LIBC`
  That are Kconfig symbols and can be used to ensure that only a valid C library is chosen for the toolchain.
- Consider deprecating the `TOOLCHAIN_HAS_NEWLIB` and migrate this to use the Kconfig knowledge.
   (outside the scope of this PR)
   However, before deprecating this `TOOLCHAIN_HAS_NEWLIB` we should be careful, cause we still have a desire to support using CMake toolchain files, so it could still have some value. (Using a CMake toolchain file should probably disable the libc choice menu completely)

I prefer to create new Kconfig settings for handling the C library support in Kconfig  instead of re-using an existing CMake symbol, but i'm open to another name than `SUPPORT_x_LIBC`.

Having the symbol `TOOLCHAIN_HAS_NEWLIB`, and the same symbol in Kconfig, which then becomes `CONFIG_TOOLCHAIN_HAS_NEWLIB` is bound to give confusion among readers.
Remember, all Kconfig symbols are pre-fixed with `CONFIG_<symbol>` when loaded into CMake.

Reply:
Ok, so the problem is that tests need to be filtered with `TOOLCHAIN_HAS_NEWLIB` and that cannot be done with Kconfig (at least with how the build process work right now).

> Consider deprecating the TOOLCHAIN_HAS_NEWLIB and migrate this to use the Kconfig knowledge.
> (outside the scope of this PR)

In that case, as you noted above, we should make `SUPPORT_NEWLIB_LIBC` derive from `TOOLCHAIN_HAS_NEWLIB`, especially if we are going to keep the CMake toolchain files (which I think we have no option but to do that). But, as noted in #12393, that is easier said than done due to how Zephyr builds and will likely require some rework in the build process.

That is certainly outside the scope of this PR. But then, having both `SUPPORT_NEWLIB_LIBC` in Kconfig and `TOOLCHAIN_HAS_NEWLIB` in CMake that are not synced to each other is even more confusing and error prone.

In that sense, should we defer adding these symbols until we have a more concrete solution? (maybe moved out to a separate PR?)

Reply:
> In that case, as you noted above, we should make SUPPORT_NEWLIB_LIBC derive from TOOLCHAIN_HAS_NEWLIB

This is actually not a bad idea. Let me try that.